### PR TITLE
[FLINK-22232][tests] Add UnalignedCheckpointsStressITCase

### PR DESCRIPTION
**Pull request by pnowojski**
This PR adds extra test coverage for unaligned checkpoints.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by dawidwys — COMMENTED** [review]
Thanks @pnowojski for the PR.

I put some suggestions to the code. The general idea seems legit ;)

**Review by AHeise — COMMENTED** [review]
I like the approach very much. Found some smaller improvements though.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+                firstCompletedCheckpoint = checkpointId;
+            }
+            if (completedCheckpointsBeforeFailure <= checkpointId - firstCompletedCheckpoint) {
+                throw new ExpectedTestException(
+                        lastProcessedRecord == null ? "no record" : lastProcessedRecord.toString());
+            }
+        }
+    }
+
+    private static class Record {
```
Comment (AHeise) [inline flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:491]: How is this record actually serialized? Kryo? I guess so
```
2021-04-16 08:11:37,984 INFO  org.apache.flink.api.java.typeutils.TypeExtractor            Class org.apache.flink.test.checkpointing.UnalignedCheckpointStressITCase$Record is not public so it cannot be used as a POJO type and must be processed as GenericType. Please read the Flink documentation on "Data Types & Serialization" for details of the effect on performance.
```

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+        checkState(matcher.find());
+        return Integer.parseInt(matcher.group(1));
+    }
+
+    private File discoverRetainedCheckpoint() throws Exception {
+        // structure: root/attempt/checkpoint/_metadata
+        File attemptDir = temporaryFolder.getRoot();
+        Path checkpointDir = null;
+
+        for (int i = 0; i <= 1000 && checkpointDir == null; i++) {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:256]: I'm not sure if @rkhachatryan would be working:
1. `Comparator.comparing(Path::toString)` I think doesn't work when comparing checkpoints with different number of digits in the checkpoint number, for example using path comparison `chk-99` > `chk-100`
2. UCRescaleITCase is relaying on cluster shutdown `miniCluster.after();` to make sure the checkpoints are flushed. I'm resubmitting new job for each iteration to the same cluster. I don't know exactly why, but I'm assuming that's why without looping and sleeps my code is failing. (on the other hand, without re-using cluster test would be slower).

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+                                .setNumberTaskManagers(NUM_TASK_MANAGERS)
+                                .setNumberSlotsPerTaskManager(NUM_TASK_SLOTS)
+                                .build());
+        cluster.before();
+
+        FileUtils.cleanDirectory(temporaryFolder.getRoot());
+    }
+
+    @After
+    public void shutDownExistingCluster() {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:127]: "shut down" are two words, so `shutDown` is the correct one, right?

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+            }
+            if (completedCheckpointsBeforeFailure <= checkpointId - firstCompletedCheckpoint) {
+                throw new ExpectedTestException(
+                        lastProcessedRecord == null ? "no record" : lastProcessedRecord.toString());
+            }
+        }
+    }
+
+    private static class Record {
+        private int sourceId;
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:492]: This was my failed first attempt  to make it PoJo. I've changed it to public non final:

> All non-static, non-transient fields in the class (and all superclasses) are either public (and non-final) or have a public getter- and a setter- method that follows the Java beans naming conventions for getters and setters.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:127]: https://www.merriam-webster.com/dictionary/shutdown

[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:127]: You might be right, however quick search through our codebase will give you a clear view that we prefer using shutdown (even in cases such as `shutdownCluster(s)`.)

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 5]
```diff
@@ -0,0 +1,553 @@
...
+            cluster.after();
+            cluster = null;
+        }
+    }
+
+    public static int getSourceId(Record record) {
+        return record.getSourceId();
+    }
+
+    public void testProgram(StreamExecutionEnvironment env) {
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:138]: Can you check the java doc if it's good enough?

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 6]
```diff
@@ -0,0 +1,553 @@
...
+            Thread.sleep(5);
+            try (Stream<Path> files = Files.walk(Paths.get(attemptDir.getPath()))) {
+                checkpointDir =
+                        files.filter(Files::isRegularFile)
+                                .filter(path -> path.endsWith("_metadata"))
+                                .map(path -> path.getParent())
+                                .sorted(
+                                        Comparator.comparingInt(
+                                                UnalignedCheckpointStressITCase
+                                                        ::getCheckpointNumberFromPath))
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:266]: Unfortunately I think yes. Removing checkpoints happens asynchronously I believe.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 7]
```diff
@@ -0,0 +1,553 @@
...
+ * A stress test that runs for a pre-defined amount of time, verifying data correctness and every
+ * couple of checkpoints is triggering fail over to stress test unaligned checkpoints.
+ */
+@SuppressWarnings("serial")
+public class UnalignedCheckpointStressITCase extends TestLogger {
+
+    protected static final int CHECKPOINT_INTERVAL = 20;
+    protected static final int MINIMUM_COMPLETED_CHECKPOINTS_BETWEEN_FAILURES = 2;
+    protected static final int MAXIMUM_COMPLETED_CHECKPOINTS_BETWEEN_FAILURES = 10;
+    protected static final long TEST_DURATION = Time.minutes(1).toMilliseconds();
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:93]: Depending on machine but it should be order of magnitude of ~20.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 8]
```diff
@@ -0,0 +1,553 @@
...
+            public int getSleep() {
+                int burstChance = 1000;
+                return ThreadLocalRandom.current().nextInt(burstChance) == 0
+                        ? burstChance * NORMAL_RECORD_SLEEP
+                        : 0;
+            }
+        };
+
+        /** @return sleep time in milliseconds */
+        public abstract int getSleep();
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:384]: I, on the other hand, am not a fan of Suppliers like that. It makes debugging and navigating IDE difficult, so I would prefer to skip this comment.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+                        payloadAt(i),
+                        i,
+                        payload[i],
+                        this);
+            }
+            return this;
+        }
+
+        private byte payloadAt(int index) {
+            return (byte) ((value + index) % 128);
```
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:527]: I'm not sure if that's worth the extra effort?

**Comment by pnowojski** [comment]
I have addressed the feedback. I have also included a couple of extra commits from @AHeise, so please pay special attention to those.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:524]: I'm also not sure. All I can say that getting any invalid payload makes it currently impossible to detect where it's coming from.

**Review by AHeise — APPROVED** [review]
LGTM.

File: flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pnowojski) [reply flink-tests/src/test/java/org/apache/flink/test/checkpointing/UnalignedCheckpointStressITCase.java:133]: `shutDown` is also being used (not that frequently, but still). As it's a non critical place (it's not an API), so consistency doesn't matter that much, I would personally prefer to keep the correct spelling here. Unless you have strong feelings in other direction?

**Pull request merged**